L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+#include <utility>
+#include <cmath>
+
+using namespace std;
+using namespace edm;
+using namespace trackerDTC;
+
+namespace trackerTFP {
+
+  KalmanFilter::KalmanFilter(const ParameterSet& iConfig, const Setup* setup, const DataFormats* dataFormats, KalmanFilterFormats* kalmanFilterFormats, int region) :
+    enableTruncation_(iConfig.getParameter<bool>("EnableTruncation")),
+    setup_(setup),
+    dataFormats_(dataFormats),
+    kalmanFilterFormats_(kalmanFilterFormats),
+    region_(region),
+    input_(dataFormats_->numChannel(Process::kf)),
+    layer_(0),
+    x0_(&kalmanFilterFormats_->format(VariableKF::x0)),
+    x1_(&kalmanFilterFormats_->format(VariableKF::x1)),
+    x2_(&kalmanFilterFormats_->format(VariableKF::x2)),
+    x3_(&kalmanFilterFormats_->format(VariableKF::x3)),
+    H00_(&kalmanFilterFormats_->format(VariableKF::H00)),
+    H12_(&kalmanFilterFormats_->format(VariableKF::H12)),
+    m0_(&kalmanFilterFormats_->format(VariableKF::m0)),
+    m1_(&kalmanFilterFormats_->format(VariableKF::m1)),
+    v0_(&kalmanFilterFormats_->format(VariableKF::v0)),
+    v1_(&kalmanFilterFormats_->format(VariableKF::v1)),
+    r0_(&kalmanFilterFormats_->format(VariableKF::r0)),
+    r1_(&kalmanFilterFormats_->format(VariableKF::r1)),
+    S00_(&kalmanFilterFormats_->format(VariableKF::S00)),
+    S01_(&kalmanFilterFormats_->format(VariableKF::S01)),
+    S12_(&kalmanFilterFormats_->format(VariableKF::S12)),
+    S13_(&kalmanFilterFormats_->format(VariableKF::S13)),
+    K00_(&kalmanFilterFormats_->format(VariableKF::K00)),
+    K10_(&kalmanFilterFormats_->format(VariableKF::K10)),
+    K21_(&kalmanFilterFormats_->format(VariableKF::K21)),
+    K31_(&kalmanFilterFormats_->format(VariableKF::K31)),
+    R00_(&kalmanFilterFormats_->format(VariableKF::R00)),
+    R11_(&kalmanFilterFormats_->format(VariableKF::R11)),
+    R00Rough_(&kalmanFilterFormats_->format(VariableKF::R00Rough)),
+    R11Rough_(&kalmanFilterFormats_->format(VariableKF::R11Rough)),
+    invR00Approx_(&kalmanFilterFormats_->format(VariableKF::invR00Approx)),
+    invR11Approx_(&kalmanFilterFormats_->format(VariableKF::invR11Approx)),
+    invR00Cor_(&kalmanFilterFormats_->format(VariableKF::invR00Cor)),
+    invR11Cor_(&kalmanFilterFormats_->format(VariableKF::invR11Cor)),
+    invR00_(&kalmanFilterFormats_->format(VariableKF::invR00)),
+    invR11_(&kalmanFilterFormats_->format(VariableKF::invR11)),
+    C00_(&kalmanFilterFormats_->format(VariableKF::C00)),
+    C01_(&kalmanFilterFormats_->format(VariableKF::C01)),
+    C11_(&kalmanFilterFormats_->format(VariableKF::C11)),
+    C22_(&kalmanFilterFormats_->format(VariableKF::C22)),
+    C23_(&kalmanFilterFormats_->format(VariableKF::C23)),
+    C33_(&kalmanFilterFormats_->format(VariableKF::C33)) {}
+
+  // read in and organize input tracks and stubs
+  void KalmanFilter::consume(const StreamsTrack& streamsTrack, const TTDTC::Streams& streamsStub) {
+    auto valid = [](const auto& frame){ return frame.first.isNonnull(); };
```

Comment:
Since KF assumes input stub stream has no gaps, throw fatal exception if any gaps detected.

Reply:
I realized that s/w and f/w can handle gaps between tracks.